driver/src/contracts/stablex_contract.rs

```diff
@@ -161,19 +161,21 @@ fn parse_auction_data(packed_auction_bytes: Vec<u8>, index: U256) -> (AccountSta
     (account_state, relevant_orders)
 }
 
-fn encode_prices_for_contract(price_vector: Vec<u128>) -> (Vec<U128>, Vec<u64>) {
-    // Representing the solution's price vector more compactly as:
-    // sorted_touched_token_ids, non_zero_prices which are logically bound by index.
-    // Example solution.prices = [3, 0, 1] will be transformed into [0, 2], [3, 1]
-    let mut prices = vec![];
-    let mut ordered_token_ids = vec![];
-    for (token_id, price) in price_vector.into_iter().enumerate().skip(1) {
-        if price > 0 {
-            prices.push(U128::from(price));
-            ordered_token_ids.push(token_id as u64);
-        }
-    }
-    (prices, ordered_token_ids)
+fn encode_prices_for_contract(price_map: HashMap<u16, u128>) -> (Vec<U128>, Vec<u64>) {
```

Comment:
nit: price_map should be taken by ref?

Reply:
Sure that sounds pretty good. How do you know when to use ref and when not?

Reply:
That’s a hard one to explain. It’s to do with the semantics of the function. Here, your function just needs to read the data to produce the output so you should use a borrow `&`. If a function needs to modify its inputs (for example, imagine a function that appends to a `String` parameter) then you need a mutable borrow `&mut`. Now, if the function needs to take the input in a way that the caller can never use it again (some examples are if you want to express that the function inputs become invalid after the function, or you are moving the input into a struct or something) then you to transfer ownership, AKA move the value into the function.

Think of them as different ways to express your functions intentions on what it will do with the data:
- `&data` just look at the data
- `&mut data` modify the data
- `data` take the data away

The reason for these it allows you to write some pretty efficient code, and choosing which one of these is usually about avoiding having multiple copies of the same data.

Reply:
That’s a hard one to explain. It’s to do with the semantics of the function. Here, your function just needs to read the data to produce the output so you should use a borrow `&`. If a function needs to modify its inputs (for example, imagine a function that appends to a `String` parameter) then you need a mutable borrow `&mut`. Now, if the function needs to take the input in a way that the caller can never use it again (some examples are if you want to express that the function inputs become invalid after the function, or you are moving the input into a struct or something) then you to transfer ownership, AKA move the value into the function.

Think of them as different ways to express your functions intentions on what it will do with the data:
- `&data` just look at the data
- `&mut data` modify the data
- `data` take the data away

The reason for these it allows you to write some pretty efficient code, and choosing which one of these is usually about avoiding having multiple copies of the same data.

Reply:
That’s a hard one to explain. It’s to do with the semantics of the function. Here, your function just needs to read the data to produce the output so you should use a borrow `&`. If a function needs to modify its inputs (for example, imagine a function that appends to a `String` parameter) then you need a mutable borrow `&mut`. Now, if the function needs to take the input in a way that the caller can never use it again (some examples are if you want to express that the function inputs become invalid after the function, or you are moving the input into a struct or something) then you to transfer ownership, AKA move the value into the function.

Think of them as different ways to express your functions intentions on what it will do with the data:
- `&data` just look at the data
- `&mut data` modify the data
- `data` take the data away

The reason for these it allows you to write some pretty efficient code, and choosing which one of these is usually about avoiding having multiple copies of the same data.